## 1. Setup


### Import

Before moving on to learn pandas first we need to install it and import it. If you install [Anaconda distributions](https://www.anaconda.com/) on your local machine or using [Google Colab](https://research.google.com/colaboratory) then pandas will already be available there, otherwise, you follow this installation process from [pandas official's website](https://pandas.pydata.org/docs/getting_started/install.html).

In [1]:
# Importing libraries
import numpy as np
import pandas as pd

In [4]:
# we can set numbers for how many rows and columns will be displayed
pd.set_option('display.min_rows', 10) #default will be 10
pd.set_option('display.max_columns', 20)

## 2. Loading Different Data Formats Into a Pandas Data Frame




### Read CSV file from URL

In [5]:
# Read csv file from url
url="https://github.com/MaskiVal/Html/raw/main/online_store_customer_data.csv"
df = pd.read_csv(url)
df.head(3)

Transaction_date  Transaction_ID  Gender   Age Marital_status State_names  \
0         1/1/2019          151200  Female  19.0         Single      Kansas   
1         1/1/2019          151201    Male  49.0         Single    Illinois   
2         1/1/2019          151202    Male  63.0        Married  New Mexico   

  Segment Employees_status Payment_method  Referal  Amount_spent  
0   Basic     Unemployment          Other      1.0       2051.36  
1   Basic    self-employed           Card      0.0        544.04  
2   Basic          workers         PayPal      1.0       1572.60

## 3. Data preprocessing
Data preprocessing is the process of making raw data to clean data. This is the most crucial part of data the science. In this section, we will explore data first then we remove unwanted columns, remove duplicates, handle missing data, etc. After this step, we get clean data from raw data.

### 3.1 Data Exploring

#### Retrieving rows from data frame.

In [6]:
# display first 3 rows
df.head(3)


Transaction_date  Transaction_ID  Gender   Age Marital_status State_names  \
0         1/1/2019          151200  Female  19.0         Single      Kansas   
1         1/1/2019          151201    Male  49.0         Single    Illinois   
2         1/1/2019          151202    Male  63.0        Married  New Mexico   

  Segment Employees_status Payment_method  Referal  Amount_spent  
0   Basic     Unemployment          Other      1.0       2051.36  
1   Basic    self-employed           Card      0.0        544.04  
2   Basic          workers         PayPal      1.0       1572.60

In [7]:
# display last 6 rows
df.tail(6)

Transaction_date  Transaction_ID  Gender   Age Marital_status  \
2506        4/30/2021          153694    Male  34.0         Single   
2507         5/1/2021          153695  Female  57.0         Single   
2508         5/1/2021          153696  Female  36.0        Married   
2509         5/1/2021          153697    Male  22.0         Single   
2510         5/1/2021          153698     NaN  44.0         Single   
2511         5/1/2021          153699    Male  48.0         Single   

         State_names   Segment Employees_status Payment_method  Referal  \
2506         Florida   Missing        Employees          Other      1.0   
2507  South Carolina  Platinum    self-employed           Card      0.0   
2508          Hawaii    Silver    self-employed         PayPal      1.0   
2509  South Carolina     Basic          workers         PayPal      1.0   
2510        New York     Basic        Employees         PayPal      0.0   
2511      California    Silver          workers         PayPal      1.0   

      Amount_spent  
2506        286.82  
2507        150.10  
2508        708.88  
2509       2030.07  
2510       1909.77  
2511       1073.15

#### Retrieving sample rows from data frame.



In [8]:
# Display random 7 sample rows
df.sample(7)

Transaction_date  Transaction_ID  Gender   Age Marital_status  \
1679        7/23/2020          152867  Female  31.0        Married   
1710         8/4/2020          152898    Male  27.0        Married   
1985       10/26/2020          153173  Female  43.0        Married   
329         4/21/2019          151517    Male  54.0         Single   
878        10/14/2019          152066    Male  26.0        Married   
2330         3/3/2021          153518  Female  73.0         Single   
1401         4/9/2020          152589  Female  19.0         Single   

         State_names   Segment Employees_status Payment_method  Referal  \
1679  South Carolina     Basic    self-employed          Other      1.0   
1710        Illinois   Missing    self-employed         PayPal      0.0   
1985    Rhode Island     Basic        Employees         PayPal      1.0   
329         New York  Platinum        Employees          Other      0.0   
878          Florida   Missing          workers         PayPal      0.0   
2330    Rhode Island     Basic    self-employed          Other      0.0   
1401         Montana      Gold        Employees         PayPal      1.0   

      Amount_spent  
1679       2778.69  
1710       1020.51  
1985           NaN  
329        2463.70  
878        2705.36  
2330       1623.74  
1401        214.11

#### Retrieving information about dataframe

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2512 entries, 0 to 2511
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction_date  2512 non-null   object 
 1   Transaction_ID    2512 non-null   int64  
 2   Gender            2484 non-null   object 
 3   Age               2470 non-null   float64
 4   Marital_status    2512 non-null   object 
 5   State_names       2512 non-null   object 
 6   Segment           2512 non-null   object 
 7   Employees_status  2486 non-null   object 
 8   Payment_method    2512 non-null   object 
 9   Referal           2357 non-null   float64
 10  Amount_spent      2270 non-null   float64
dtypes: float64(3), int64(1), object(7)
memory usage: 216.0+ KB


In [10]:
# display datatypes
df.dtypes

Transaction_date     object
Transaction_ID        int64
Gender               object
Age                 float64
Marital_status       object
State_names          object
Segment              object
Employees_status     object
Payment_method       object
Referal             float64
Amount_spent        float64
dtype: object

#### Display number of rows and columns.

In [11]:
df.shape

(2512, 11)

In [21]:
df.columns

Index(['Transaction_date', 'Transaction_ID', 'Gender', 'Age', 'Marital_status',
       'State_names', 'Segment', 'Employees_status', 'Payment_method',
       'Referal', 'Amount_spent'],
      dtype='object')

In [25]:
df.loc[:2,'Age']

0    19.0
1    49.0
2    63.0
Name: Age, dtype: float64

In [28]:
df[['Age','Transaction_date','Gender']].iloc[:4,::]

Age Transaction_date  Gender
0  19.0         1/1/2019  Female
1  49.0         1/1/2019    Male
2  63.0         1/1/2019    Male
3  18.0         1/1/2019     NaN

#### Retrieving a Range of Rows

In [37]:
# for display 2nd to 6th rows
df.iloc[1:6,::]
# for display starting to 10th
df.iloc[9::,::] 

# for display last two rows
df.iloc[-2::,::]
 


Transaction_date  Transaction_ID Gender   Age Marital_status State_names  \
2510         5/1/2021          153698    NaN  44.0         Single    New York   
2511         5/1/2021          153699   Male  48.0         Single  California   

     Segment Employees_status Payment_method  Referal  Amount_spent  
2510   Basic        Employees         PayPal      0.0       1909.77  
2511  Silver          workers         PayPal      1.0       1073.15

 ### 3.2 Data Cleaning
After the explore our datasets may need to clean them for better analysis. Data coming in from multiple sources so It's possible to have an error in some values. This is where data cleaning becomes extremely important. In this section, we will delete unwanted columns, rename columns, correct appropriate data types, etc.


#### Delete Columns name

In [11]:
# Drop unwanted columns 'Transaction_ID'

df.drop(columns= 'Transaction_ID')

Transaction_date  Gender   Age Marital_status     State_names   Segment  \
0            1/1/2019  Female  19.0         Single          Kansas     Basic   
1            1/1/2019    Male  49.0         Single        Illinois     Basic   
2            1/1/2019    Male  63.0        Married      New Mexico     Basic   
3            1/1/2019     NaN  18.0         Single        Virginia  Platinum   
4            1/1/2019    Male  27.0         Single     Connecticut     Basic   
...               ...     ...   ...            ...             ...       ...   
2507         5/1/2021  Female  57.0         Single  South Carolina  Platinum   
2508         5/1/2021  Female  36.0        Married          Hawaii    Silver   
2509         5/1/2021    Male  22.0         Single  South Carolina     Basic   
2510         5/1/2021     NaN  44.0         Single        New York     Basic   
2511         5/1/2021    Male  48.0         Single      California    Silver   

     Employees_status Payment_method  Referal  Amount_spent  
0        Unemployment          Other      1.0       2051.36  
1       self-employed           Card      0.0        544.04  
2             workers         PayPal      1.0       1572.60  
3             workers           Card      1.0       1199.79  
4       self-employed           Card      0.0           NaN  
...               ...            ...      ...           ...  
2507    self-employed           Card      0.0        150.10  
2508    self-employed         PayPal      1.0        708.88  
2509          workers         PayPal      1.0       2030.07  
2510        Employees         PayPal      0.0       1909.77  
2511          workers         PayPal      1.0       1073.15  

[2512 rows x 10 columns]

#### Change Columns name

In [12]:
df_col = df.copy()
df_col.rename( columns= {'Gender':'Sex','Transaction_date': 'Date' },inplace= True)

#### Adding a new column to a DataFrame



In [28]:
# Add a new ajusted column which value will be Amount_spent * 100
df['new_col'] = df['Amount_spent'].apply(lambda x: x*10)
df['new_col']

0       20513.6
1        5440.4
2       15726.0
3       11997.9
4           NaN
         ...   
2507     1501.0
2508     7088.8
2509    20300.7
2510    19097.7
2511    10731.5
Name: new_col, Length: 2512, dtype: float64

#### String value change or replace

In [ ]:
df_col.head(3)

Date  Transaction_ID     Sex   Age Marital_status State_names Segment  \
0  1/1/2019          151200  Female  19.0         Single      Kansas   Basic   
1  1/1/2019          151201    Male  49.0         Single    Illinois   Basic   
2  1/1/2019          151202    Male  63.0        Married  New Mexico   Basic   

  Employees_status Payment_method  Referal  Amount_spent   new_col  
0     Unemployment          Other      1.0       2051.36  205136.0  
1    self-employed           Card      0.0        544.04   54404.0  
2          workers         PayPal      1.0       1572.60  157260.0

In [16]:
df_col['Sex'] = df_col['Sex'].str.replace("Female", "Woman")
df_col['Sex'] = df_col['Sex'].str.replace("Male", "Man")
df_col['Sex'] 

0       Woman
1         Man
2         Man
3         NaN
4         Man
        ...  
2507    Woman
2508    Woman
2509      Man
2510      NaN
2511      Man
Name: Sex, Length: 2512, dtype: object

In [17]:
df_col['Sex'].isna().sum()

28

In [18]:
# changing Female to Woman and Male to Man in Sex column.
#first argument in loc function is condition and second one is columns name.

df_col['Sex']



0       Woman
1         Man
2         Man
3         NaN
4         Man
        ...  
2507    Woman
2508    Woman
2509      Man
2510      NaN
2511      Man
Name: Sex, Length: 2512, dtype: object

In [54]:
df_col.head(3)

Transaction_date  Transaction_ID    Sex   Age Marital_status State_names  \
0         1/1/2019          151200  Woman  19.0         Single      Kansas   
1         1/1/2019          151201    Man  49.0         Single    Illinois   
2         1/1/2019          151202    Man  63.0        Married  New Mexico   

  Segment Employees_status Payment_method  Referal  Amount_spent      New  \
0   Basic     Unemployment          Other      1.0       2051.36  20513.6   
1   Basic    self-employed           Card      0.0        544.04   5440.4   
2   Basic          workers         PayPal      1.0       1572.60  15726.0   

   new_col  
0  20513.6  
1   5440.4  
2  15726.0

Now Sex columns values are changed Female to Woman and Male to Man.

#### Datatypes change

In [60]:
df_col.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2512 entries, 0 to 2511
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date              2512 non-null   object 
 1   Transaction_ID    2512 non-null   int64  
 2   Sex               2484 non-null   object 
 3   Age               2470 non-null   float64
 4   Marital_status    2512 non-null   object 
 5   State_names       2512 non-null   object 
 6   Segment           2512 non-null   object 
 7   Employees_status  2486 non-null   object 
 8   Payment_method    2512 non-null   object 
 9   Referal           2357 non-null   float64
 10  Amount_spent      2270 non-null   float64
 11  New               2270 non-null   float64
 12  new_col           2270 non-null   float64
dtypes: float64(5), int64(1), object(7)
memory usage: 255.3+ KB


In our `Date` columns, it's object type so now we will convert this to date types, and also we will convert `Referal` columns float64 to float32.

In [19]:
# change object type to datetime64 format
df_col['Date'] = df_col['Date'].astype("datetime64[ns]")

# change float64 to float32 of Referal columns
df_col["Referal"] = df_col["Referal"].astype("float32")


In [20]:
df_col.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2512 entries, 0 to 2511
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Date              2512 non-null   datetime64[ns]
 1   Transaction_ID    2512 non-null   int64         
 2   Sex               2484 non-null   object        
 3   Age               2470 non-null   float64       
 4   Marital_status    2512 non-null   object        
 5   State_names       2512 non-null   object        
 6   Segment           2512 non-null   object        
 7   Employees_status  2486 non-null   object        
 8   Payment_method    2512 non-null   object        
 9   Referal           2357 non-null   float32       
 10  Amount_spent      2270 non-null   float64       
dtypes: datetime64[ns](1), float32(1), float64(2), int64(1), object(6)
memory usage: 206.2+ KB


### 3.3 Remove duplicate

In [21]:
# Display duplicated entries
df_col.duplicated().value_counts()

False    2500
True       12
Name: count, dtype: int64

In [22]:
# duplicate rows dispaly, keep arguments will--- 'first', 'last' and False
duplicate_value = df_col.duplicated()

df_col.loc[duplicate_value, :]

Date  Transaction_ID    Sex   Age Marital_status     State_names  \
64  2019-01-25          151259    Man  73.0        Married   West Virginia   
65  2019-01-26          151260    Man  55.0        Married          Kansas   
66  2019-01-26          151261  Woman  72.0        Married            Iowa   
67  2019-01-26          151262    Man  15.0        Married  South Carolina   
68  2019-01-27          151263  Woman  63.0         Single           Texas   
109 2019-02-06          151297    Man  60.0        Married            Utah   
110 2019-02-07          151298  Woman  45.0        Married        Missouri   
111 2019-02-08          151299    Man  33.0         Single         Arizona   
112 2019-02-08          151300    Man  24.0        Married  South Carolina   
113 2019-02-08          151301  Woman  53.0         Single        Colorado   
114 2019-02-08          151302  Woman  70.0        Married   New Hampshire   
115 2019-02-08          151303    Man  16.0        Married        Nebraska   

      Segment Employees_status Payment_method  Referal  Amount_spent  
64      Basic        Employees         PayPal      0.0       1397.09  
65      Basic        Employees          Other      1.0       1277.64  
66     Silver     Unemployment         PayPal      NaN        515.77  
67      Basic    self-employed          Other      1.0        790.10  
68       Gold        Employees           Card      1.0       1218.56  
109    Silver     Unemployment          Other      1.0        433.20  
110  Platinum          workers          Other      1.0        929.89  
111    Silver          workers         PayPal      0.0       2560.26  
112     Basic     Unemployment          Other      0.0           NaN  
113     Basic    self-employed          Other      1.0       1888.69  
114   Missing        Employees           Card      1.0        685.49  
115     Basic          workers         PayPal      1.0       1264.73

In [23]:
# dropping ALL duplicate values
df_col.drop_duplicates()
df_col

Date  Transaction_ID    Sex   Age Marital_status     State_names  \
0    2019-01-01          151200  Woman  19.0         Single          Kansas   
1    2019-01-01          151201    Man  49.0         Single        Illinois   
2    2019-01-01          151202    Man  63.0        Married      New Mexico   
3    2019-01-01          151203    NaN  18.0         Single        Virginia   
4    2019-01-01          151204    Man  27.0         Single     Connecticut   
...         ...             ...    ...   ...            ...             ...   
2507 2021-05-01          153695  Woman  57.0         Single  South Carolina   
2508 2021-05-01          153696  Woman  36.0        Married          Hawaii   
2509 2021-05-01          153697    Man  22.0         Single  South Carolina   
2510 2021-05-01          153698    NaN  44.0         Single        New York   
2511 2021-05-01          153699    Man  48.0         Single      California   

       Segment Employees_status Payment_method  Referal  Amount_spent  
0        Basic     Unemployment          Other      1.0       2051.36  
1        Basic    self-employed           Card      0.0        544.04  
2        Basic          workers         PayPal      1.0       1572.60  
3     Platinum          workers           Card      1.0       1199.79  
4        Basic    self-employed           Card      0.0           NaN  
...        ...              ...            ...      ...           ...  
2507  Platinum    self-employed           Card      0.0        150.10  
2508    Silver    self-employed         PayPal      1.0        708.88  
2509     Basic          workers         PayPal      1.0       2030.07  
2510     Basic        Employees         PayPal      0.0       1909.77  
2511    Silver          workers         PayPal      1.0       1073.15  

[2512 rows x 11 columns]

### 3.4 Handling missing values

Handling missing values in the common task in the data pre-processing part. For many reasons most of the time we will encounter missing values. Without dealing with this we can't do the proper model building. For this section first, we will find out missing values then we decided how to handle them. We can handle this by removing affected columns or rows or replacing appropriate values there.

#### Display missing values information

In [24]:
df_col.isna().sum()

Date                  0
Transaction_ID        0
Sex                  28
Age                  42
Marital_status        0
State_names           0
Segment               0
Employees_status     26
Payment_method        0
Referal             155
Amount_spent        242
dtype: int64

#### Delete Nan rows

If we have less Nan value then we can delete entire rows by `dropna()` function. For this function, we will add columns name in subset parameter.

In [25]:
# df copy to df_copy
df_col_copy = df_col.copy()

In [26]:
#Delete Nan rows of Job Columns

df_col_copy.isna().sum().sort_values(ascending = False)

Amount_spent        242
Referal             155
Age                  42
Sex                  28
Employees_status     26
Date                  0
Transaction_ID        0
Marital_status        0
State_names           0
Segment               0
Payment_method        0
dtype: int64

#### Delete entire columns

If we have a large number of nan values in particular columns then dropping those columns might be a good decision rather than imputing.

In [30]:
df_col_copy.drop(columns = ["Amount_spent"], inplace = True, errors="ignore")
df_col_copy.drop(columns = ["new_col"], inplace = True, errors= "ignore")


In [31]:
df_col_copy.isna().sum().sort_values(ascending = False)

Referal             155
Age                  42
Sex                  28
Employees_status     26
Date                  0
Transaction_ID        0
Marital_status        0
State_names           0
Segment               0
Payment_method        0
dtype: int64

#### Impute missing values

Sometimes if we delete entire columns that will be not the appropriate approach. Delete columns can affect our model building because we will lose our main features. For imputing we have many approaches so here are some of the most popular techniques.

**Method 1** - Impute fixed value like 0, 'Unknown' or 'Missing' etc. We inpute Unknown in Gender columns

In [32]:


df_col.Sex.value_counts()

Sex
Woman    1356
Man      1128
Name: count, dtype: int64

**Method 2** - Impute Mean, Median and Mode

In [35]:
# Impute Mean in Amount_spent columns
df_col_mean_Am_S =df_col["Amount_spent"].mean() 
df_col["Amount_spent"].fillna(df_col_mean_Am_S, inplace = True)

#Impute Median in Age column

df_col_median_age = df_col["Age"].median()
df_col["Age"].fillna(df_col_median_age, inplace = True)


# Impute Mode in Employees_status column

df_col_mode_emp_status = df_col['Employees_status'].mode()
df_col.Employees_status.fillna(df_col_mode_emp_status, inplace = True)


C:\Users\nhann\AppData\Local\Temp\ipykernel_28304\1072930563.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_col["Amount_spent"].fillna(df_col_mean_Am_S, inplace = True)
C:\Users\nhann\AppData\Local\Temp\ipykernel_28304\1072930563.py:8: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a

In [ ]:
df_col

Date  Transaction_ID      Sex   Age Marital_status     State_names  \
0    2019-01-01          151200    Woman  19.0         Single          Kansas   
1    2019-01-01          151201      Man  49.0         Single        Illinois   
2    2019-01-01          151202      Man  63.0        Married      New Mexico   
3    2019-01-01          151203  Unknown  18.0         Single        Virginia   
4    2019-01-01          151204      Man  27.0         Single     Connecticut   
...         ...             ...      ...   ...            ...             ...   
2507 2021-05-01          153695    Woman  57.0         Single  South Carolina   
2508 2021-05-01          153696    Woman  36.0        Married          Hawaii   
2509 2021-05-01          153697      Man  22.0         Single  South Carolina   
2510 2021-05-01          153698  Unknown  44.0         Single        New York   
2511 2021-05-01          153699      Man  48.0         Single      California   

       Segment Employees_status Payment_method  Referal  Amount_spent  \
0        Basic     Unemployment          Other      1.0    2051.36000   
1        Basic    self-employed           Card      0.0     544.04000   
2        Basic          workers         PayPal      1.0    1572.60000   
3     Platinum          workers           Card      1.0    1199.79000   
4        Basic    self-employed           Card      0.0    1419.59178   
...        ...              ...            ...      ...           ...   
2507  Platinum    self-employed           Card      0.0     150.10000   
2508    Silver    self-employed         PayPal      1.0     708.88000   
2509     Basic          workers         PayPal      1.0    2030.07000   
2510     Basic        Employees         PayPal      0.0    1909.77000   
2511    Silver          workers         PayPal      1.0    1073.15000   

       new_col  
0     205136.0  
1      54404.0  
2     157260.0  
3     119979.0  
4          NaN  
...        ...  
2507   15010.0  
2508   70888.0  
2509  203007.0  
2510  190977.0  
2511  107315.0  

[2500 rows x 12 columns]

**Method 3** - Imputing forward fill or backfill by `ffill` and `bfill`. In `ffill` missing value impute from the value of the above row and for `bfill` it's taken from the below rows value.

In [36]:
df_col.Referal.fillna(method = "ffill", inplace = True)

C:\Users\nhann\AppData\Local\Temp\ipykernel_28304\291345715.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_col.Referal.fillna(method = "ffill", inplace = True)
C:\Users\nhann\AppData\Local\Temp\ipykernel_28304\291345715.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_col.Referal.fillna(method = "ffill", inplace = True)


In [ ]:

df_col.isna().sum().sum()

54

Now we deal with all missing values with different methods. So now we haven't any null values.

## 4. Memory management

When we work on large datasets, There we get one big issue is a memory problem. We need too large resources for dealing with this. But there are some methods in pandas to deal with this. Here are some methods or strategies to deal with this problem with help of pandas.

### Change datatypes

In [39]:
df_memory = df_col.copy()

In [40]:
memory_usage = df_memory.memory_usage(deep=True)
memory_usage_in_mbs = round(np.sum(memory_usage / 1024 ** 2), 3)
print(f" Total memory taking df_memory dataframe is : {memory_usage_in_mbs:.2f} MB ")

 Total memory taking df_memory dataframe is : 1.00 MB 


#### Change object to category datatypes

Our data frame is small in size. Which is 1 MB. Now We will convert our object datatype to category.

In [41]:
# Object datatype to category convert
df_memory[df_memory.select_dtypes(["object"]).columns]

Sex Marital_status     State_names   Segment Employees_status  \
0     Woman         Single          Kansas     Basic     Unemployment   
1       Man         Single        Illinois     Basic    self-employed   
2       Man        Married      New Mexico     Basic          workers   
3       NaN         Single        Virginia  Platinum          workers   
4       Man         Single     Connecticut     Basic    self-employed   
...     ...            ...             ...       ...              ...   
2507  Woman         Single  South Carolina  Platinum    self-employed   
2508  Woman        Married          Hawaii    Silver    self-employed   
2509    Man         Single  South Carolina     Basic          workers   
2510    NaN         Single        New York     Basic        Employees   
2511    Man         Single      California    Silver          workers   

     Payment_method  
0             Other  
1              Card  
2            PayPal  
3              Card  
4              Card  
...             ...  
2507           Card  
2508         PayPal  
2509         PayPal  
2510         PayPal  
2511         PayPal  

[2512 rows x 6 columns]

In [42]:
# convert object to category
df_memory.info(memory_usage = "deep")


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2512 entries, 0 to 2511
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Date              2512 non-null   datetime64[ns]
 1   Transaction_ID    2512 non-null   int64         
 2   Sex               2484 non-null   object        
 3   Age               2512 non-null   float64       
 4   Marital_status    2512 non-null   object        
 5   State_names       2512 non-null   object        
 6   Segment           2512 non-null   object        
 7   Employees_status  2486 non-null   object        
 8   Payment_method    2512 non-null   object        
 9   Referal           2512 non-null   float32       
 10  Amount_spent      2512 non-null   float64       
dtypes: datetime64[ns](1), float32(1), float64(2), int64(1), object(6)
memory usage: 1022.7 KB


Now its reduce 1.15 megabytes to 216.6 kb. It's almost reduced 5.5 times.

#### Change int64 or float64 to int 32, 16, or 8

By default, pandas store numeric values to int64 or float64. Which takes more memory. If we have to store small numbers then we can change to 64 to 32, 16, and so on. For example, our Referal columns have only 0 and 1 values so for that we don't need to store at float64. so now we change it to float16.

In [43]:
# Change Referal column datatypes
df_memory['Age'] = df_memory['Age'].astype('float32')


In [44]:
# convert object to category
df_memory.info(memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2512 entries, 0 to 2511
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Date              2512 non-null   datetime64[ns]
 1   Transaction_ID    2512 non-null   int64         
 2   Sex               2484 non-null   object        
 3   Age               2512 non-null   float32       
 4   Marital_status    2512 non-null   object        
 5   State_names       2512 non-null   object        
 6   Segment           2512 non-null   object        
 7   Employees_status  2486 non-null   object        
 8   Payment_method    2512 non-null   object        
 9   Referal           2512 non-null   float32       
 10  Amount_spent      2512 non-null   float64       
dtypes: datetime64[ns](1), float32(2), float64(1), int64(1), object(6)
memory usage: 1012.9 KB


After changing only one column's data types we reduce 216 kb to 179 kb.


**Note: Before changing datatypes please make sure it's consequences.**

## 5. Data Analysis

### 5.1. Calculating Basic statistical measurement

In [45]:
df_col.describe().T

count                           mean                  min  \
Date              2512  2020-02-25 06:14:19.872611328  2019-01-01 00:00:00   
Transaction_ID  2512.0                  152443.931131             151200.0   
Age             2512.0                       46.64371                 15.0   
Referal         2512.0                       0.655255                  0.0   
Amount_spent    2512.0                    1418.422577                 2.09   

                                25%                  50%                  75%  \
Date            2019-07-23 00:00:00  2020-02-16 00:00:00  2020-09-23 00:00:00   
Transaction_ID            151815.75             152443.5            153071.25   
Age                            32.0                 47.0                 62.0   
Referal                         0.0                  1.0                  1.0   
Amount_spent               734.8625          1418.422577            1955.9975   

                                max         std  
Date            2021-05-01 00:00:00         NaN  
Transaction_ID             153699.0  724.580482  
Age                            78.0     18.0336  
Referal                         1.0     0.47538  
Amount_spent                2999.98  835.101648

We know already above code will display only numeric columns basic statistical information. for object or category columns we can use `describe(include=object)` .

In [48]:
df_col.describe(include = object).T

count unique        top  freq
Sex               2484      2      Woman  1356
Marital_status    2512      2    Married  1473
State_names       2512     50   Illinois    67
Segment           2512      5      Basic  1136
Employees_status  2486      4  Employees   946
Payment_method    2512      3     PayPal  1168

We can calculate the mean, median, mode, maximum values, minimum values of individual columns we simply use these functions.

In [50]:
# Calculate Mean
mean = df_col['Age'].mean()

# Calculate Median
median = df_col['Age'].median()

#Calculate Mode
mode = df_col['Age'].mode()

# Calculate standard deviation
std =  df_col['Age'].std()

# Calculate Minimum values
min_val = df_col['Age'].min()

# Calculate Maximum values
max_cal =df_col['Age'].max()

print(f" Mean of Age : {mean}")
print(f" Median of Age : {median}")
print(f" Mode of Age : {mode}")
print(f" Standard deviation of Age : {std:.2f}")
print(f" Maximum of Age : {max_cal}")
print(f" Menimum of Age : {min_val}")

 Mean of Age : 46.6437101910828
 Median of Age : 47.0
 Mode of Age : 0    47.0
Name: Age, dtype: float64
 Standard deviation of Age : 18.03
 Maximum of Age : 78.0
 Menimum of Age : 15.0


In pandas we can display the correlation of different numeric columns. For this we can use `.corr()` function.

In [51]:
# calculate correlation
df_col.select_dtypes('number').corr()

Transaction_ID       Age   Referal  Amount_spent
Transaction_ID        1.000000 -0.008881 -0.008488     -0.006338
Age                  -0.008881  1.000000  0.012750     -0.022128
Referal              -0.008488  0.012750  1.000000      0.000292
Amount_spent         -0.006338 -0.022128  0.000292      1.000000

### 5.2 Basic built in function for data analysis

#### Number of uniqe values in category column

In [52]:
# for display how many unique values are there in State_names column
df_col['State_names'].nunique()

50

#### Shows all unique values

In [53]:
# for display uniqe values of State_names column
df_col['State_names'].unique()

array(['Kansas', 'Illinois', 'New Mexico', 'Virginia', 'Connecticut',
       'Hawaii', 'Florida', 'Vermont', 'California', 'Colorado', 'Iowa',
       'South Carolina', 'New York', 'Maine', 'Maryland', 'Missouri',
       'North Dakota', 'Ohio', 'Nebraska', 'Montana', 'Indiana',
       'Wisconsin', 'Alabama', 'Arkansas', 'Pennsylvania',
       'New Hampshire', 'Washington', 'Texas', 'Kentucky',
       'Massachusetts', 'Wyoming', 'Louisiana', 'North Carolina',
       'Rhode Island', 'West Virginia', 'Tennessee', 'Oregon', 'Alaska',
       'Oklahoma', 'Nevada', 'New Jersey', 'Michigan', 'Utah', 'Arizona',
       'South Dakota', 'Georgia', 'Idaho', 'Mississippi', 'Minnesota',
       'Delaware'], dtype=object)

#### Counts of unique values

In [54]:
df_col['Sex'].value_counts()

Sex
Woman    1356
Man      1128
Name: count, dtype: int64

If we want to show with the percentage of occurrence rather number than we use `normalize=True` argument in `value_counts()` function

In [55]:
# Calculate percentage of each category
df_col['Sex'].value_counts(normalize= True)

Sex
Woman    0.545894
Man      0.454106
Name: proportion, dtype: float64

In [56]:
df_col['State_names'].value_counts()

State_names
Illinois          67
Georgia           64
Massachusetts     63
Maine             62
Kentucky          59
Minnesota         59
Missouri          57
Arizona           56
Delaware          56
Arkansas          55
Nevada            55
New York          55
California        55
New Mexico        55
Vermont           54
West Virginia     54
New Jersey        53
Oregon            53
Florida           53
Connecticut       52
Washington        52
Pennsylvania      52
Colorado          52
South Dakota      52
Michigan          52
Iowa              51
Rhode Island      51
Louisiana         49
Texas             49
Montana           48
North Carolina    46
Virginia          46
Hawaii            46
Wisconsin         45
Idaho             45
Maryland          45
South Carolina    45
Wyoming           44
Mississippi       44
Oklahoma          44
Utah              43
New Hampshire     43
Indiana           43
Ohio              43
Alabama           42
Alaska            41
Nebraska          41
N

In [57]:
# Sort Values by State_names
df_col.sort_values(by= 'State_names')

Date  Transaction_ID    Sex   Age Marital_status State_names  \
617  2019-07-21          151805    Man  43.0        Married     Alabama   
1856 2020-09-15          153044  Woman  18.0        Married     Alabama   
2443 2021-04-09          153631    Man  60.0        Married     Alabama   
2441 2021-04-09          153629    Man  18.0         Single     Alabama   
1023 2019-12-03          152211  Woman  30.0        Married     Alabama   
...         ...             ...    ...   ...            ...         ...   
1789 2020-08-26          152977  Woman  36.0        Married     Wyoming   
1811 2020-08-30          152999    Man  71.0        Married     Wyoming   
1819 2020-09-03          153007    Man  35.0        Married     Wyoming   
1844 2020-09-09          153032  Woman  24.0        Married     Wyoming   
119  2019-02-11          151307  Woman  43.0        Married     Wyoming   

       Segment Employees_status Payment_method  Referal  Amount_spent  
617     Silver        Employees         PayPal      1.0     87.880000  
1856    Silver          workers         PayPal      0.0   2091.730000  
2443  Platinum        Employees         PayPal      1.0   1418.422577  
2441  Platinum        Employees           Card      0.0    289.850000  
1023  Platinum     Unemployment          Other      0.0   1418.422577  
...        ...              ...            ...      ...           ...  
1789    Silver        Employees         PayPal      1.0    698.920000  
1811     Basic     Unemployment         PayPal      0.0   1647.300000  
1819     Basic          workers           Card      0.0    984.700000  
1844     Basic     Unemployment         PayPal      1.0     16.270000  
119      Basic        Employees           Card      1.0    914.940000  

[2512 rows x 11 columns]

For sorting our dataframe by Amount_spent with ascending order:

In [58]:
# Sort Values Amount_spent with ascending order
df_col.sort_values(by= 'Amount_spent', ascending= True)

Date  Transaction_ID    Sex   Age Marital_status     State_names  \
2468 2021-04-18          153656  Woman  73.0        Married        Michigan   
568  2019-07-04          151756    Man  46.0         Single  South Carolina   
2401 2021-03-25          153589  Woman  60.0         Single        Maryland   
962  2019-11-13          152150  Woman  56.0        Married          Nevada   
860  2019-10-10          152048    Man  26.0         Single        Maryland   
...         ...             ...    ...   ...            ...             ...   
743  2019-08-30          151931    Man  44.0         Single        Illinois   
589  2019-07-13          151777    Man  51.0         Single  North Carolina   
2279 2021-02-15          153467  Woman  78.0         Single         Arizona   
485  2019-06-07          151673    Man  65.0        Married         Arizona   
17   2019-01-07          151217  Woman  77.0        Married      New Mexico   

       Segment Employees_status Payment_method  Referal  Amount_spent  
2468      Gold        Employees         PayPal      1.0          2.09  
568       Gold          workers         PayPal      0.0          2.16  
2401    Silver        Employees         PayPal      1.0          2.84  
962      Basic          workers          Other      1.0          5.31  
860     Silver     Unemployment         PayPal      1.0          5.55  
...        ...              ...            ...      ...           ...  
743      Basic        Employees           Card      1.0       2996.82  
589    Missing        Employees         PayPal      1.0       2997.15  
2279    Silver        Employees         PayPal      1.0       2997.21  
485       Gold    self-employed         PayPal      1.0       2998.62  
17    Platinum          workers           Card      0.0       2999.98  

[2512 rows x 11 columns]

For sorting our dataframe by Amount_spent with descending order:

In [61]:
# Sort Values Amount_spent with descending order
df_col.sort_values(by= 'Amount_spent', ascending= False)

Date  Transaction_ID    Sex   Age Marital_status     State_names  \
17   2019-01-07          151217  Woman  77.0        Married      New Mexico   
485  2019-06-07          151673    Man  65.0        Married         Arizona   
2279 2021-02-15          153467  Woman  78.0         Single         Arizona   
589  2019-07-13          151777    Man  51.0         Single  North Carolina   
743  2019-08-30          151931    Man  44.0         Single        Illinois   
...         ...             ...    ...   ...            ...             ...   
860  2019-10-10          152048    Man  26.0         Single        Maryland   
962  2019-11-13          152150  Woman  56.0        Married          Nevada   
2401 2021-03-25          153589  Woman  60.0         Single        Maryland   
568  2019-07-04          151756    Man  46.0         Single  South Carolina   
2468 2021-04-18          153656  Woman  73.0        Married        Michigan   

       Segment Employees_status Payment_method  Referal  Amount_spent  
17    Platinum          workers           Card      0.0       2999.98  
485       Gold    self-employed         PayPal      1.0       2998.62  
2279    Silver        Employees         PayPal      1.0       2997.21  
589    Missing        Employees         PayPal      1.0       2997.15  
743      Basic        Employees           Card      1.0       2996.82  
...        ...              ...            ...      ...           ...  
860     Silver     Unemployment         PayPal      1.0          5.55  
962      Basic          workers          Other      1.0          5.31  
2401    Silver        Employees         PayPal      1.0          2.84  
568       Gold          workers         PayPal      0.0          2.16  
2468      Gold        Employees         PayPal      1.0          2.09  

[2512 rows x 11 columns]

Alternatively, We can use `nlargest()` and `nsmallest()` functions for displaying largest and smallest values with desired numbers. for example, If we want to display 4 largest Amount_spent rows then we use this:

In [63]:
df_col.nlargest(n=4, columns = 'Amount_spent')

Date  Transaction_ID    Sex   Age Marital_status     State_names  \
17   2019-01-07          151217  Woman  77.0        Married      New Mexico   
485  2019-06-07          151673    Man  65.0        Married         Arizona   
2279 2021-02-15          153467  Woman  78.0         Single         Arizona   
589  2019-07-13          151777    Man  51.0         Single  North Carolina   

       Segment Employees_status Payment_method  Referal  Amount_spent  
17    Platinum          workers           Card      0.0       2999.98  
485       Gold    self-employed         PayPal      1.0       2998.62  
2279    Silver        Employees         PayPal      1.0       2997.21  
589    Missing        Employees         PayPal      1.0       2997.15

For 3 smallest Amount_spent rows

In [64]:
df_col.nsmallest(n=3, columns='Amount_spent')

Date  Transaction_ID    Sex   Age Marital_status     State_names  \
2468 2021-04-18          153656  Woman  73.0        Married        Michigan   
568  2019-07-04          151756    Man  46.0         Single  South Carolina   
2401 2021-03-25          153589  Woman  60.0         Single        Maryland   

     Segment Employees_status Payment_method  Referal  Amount_spent  
2468    Gold        Employees         PayPal      1.0          2.09  
568     Gold          workers         PayPal      0.0          2.16  
2401  Silver        Employees         PayPal      1.0          2.84

#### Conditional queries on Data

In [65]:
# filtering - Only show Paypal users
condition = df_col['Payment_method'] == "PayPal"
df_col[condition].head(4)

Date  Transaction_ID    Sex   Age Marital_status  State_names  \
2 2019-01-01          151202    Man  63.0        Married   New Mexico   
5 2019-01-03          151205    Man  71.0         Single       Hawaii   
6 2019-01-03          151206  Woman  34.0        Married   New Mexico   
7 2019-01-03          151207    Man  37.0        Married  Connecticut   

    Segment Employees_status Payment_method  Referal  Amount_spent  
2     Basic          workers         PayPal      1.0       1572.60  
5     Basic        Employees         PayPal      1.0       2922.66  
6  Platinum        Employees         PayPal      1.0       1481.42  
7     Basic          workers         PayPal      1.0       1149.55

We can apply multiple conditional queries like before. For example, if we want to display all Married female people who lived in New York then we use the following:

In [68]:
# first create 3 condition
female_person = df_col['Sex'] == "Woman"
married_person = df_col['Marital_status'] == "Married"
loc_newyork = df_col['State_names'] == "New York"

# we passing condition on our dataframe
df_col[female_person & married_person & loc_newyork]

Date  Transaction_ID    Sex   Age Marital_status State_names  \
164  2019-03-01          151352  Woman  64.0        Married    New York   
180  2019-03-07          151368  Woman  20.0        Married    New York   
254  2019-03-31          151442  Woman  78.0        Married    New York   
282  2019-04-08          151470  Woman  32.0        Married    New York   
827  2019-09-30          152015  Woman  77.0        Married    New York   
978  2019-11-19          152166  Woman  19.0        Married    New York   
1011 2019-11-29          152199  Woman  18.0        Married    New York   
1468 2020-05-07          152656  Woman  37.0        Married    New York   
1648 2020-07-09          152836  Woman  70.0        Married    New York   
1654 2020-07-12          152842  Woman  32.0        Married    New York   
1711 2020-08-04          152899  Woman  46.0        Married    New York   
1846 2020-09-12          153034  Woman  57.0        Married    New York   
2145 2020-12-27          153333  Woman  31.0        Married    New York   
2192 2021-01-13          153380  Woman  77.0        Married    New York   
2456 2021-04-14          153644  Woman  30.0        Married    New York   

       Segment Employees_status Payment_method  Referal  Amount_spent  
164      Basic        Employees         PayPal      1.0   1581.770000  
180      Basic          workers         PayPal      1.0   2694.200000  
254    Missing        Employees         PayPal      1.0   2959.540000  
282       Gold     Unemployment          Other      1.0    522.240000  
827   Platinum          workers           Card      1.0   1418.422577  
978      Basic        Employees           Card      0.0   2473.870000  
1011     Basic    self-employed         PayPal      1.0   1340.650000  
1468     Basic        Employees           Card      0.0   2003.850000  
1648     Basic    self-employed         PayPal      1.0   1631.460000  
1654      Gold        Employees         PayPal      1.0    270.430000  
1711  Platinum        Employees           Card      1.0   1418.422577  
1846     Basic          workers           Card      0.0   1890.940000  
2145    Silver        Employees          Other      1.0    191.210000  
2192     Basic     Unemployment         PayPal      1.0    388.480000  
2456     Basic        Employees          Other      0.0    666.290000

### 5.3 Summarizing or grouping data

#### Groupby


**Grouping by one column:** For example, if we want to find `maximum` values of `Age` and `Amount_spent` by `Gender` then we can use this:

In [77]:
df_col['Sex'].isna().sum()
df_col.fillna("Unknow", inplace= True)

In [79]:
df_col[['Age','Amount_spent','Sex']].groupby(by='Sex').agg(Age = ('Age','max'), Amount_spent = ('Amount_spent', 'max'))

Age  Amount_spent
Sex                       
Man     78.0       2998.62
Unknow  72.0       2909.85
Woman   78.0       2999.98

To find `mean`, `count`, and `max` values of `Age` and `Amount_spent` by `Gender` then we can use `agg()` function with `groupby()` .

In [113]:
df_grouped = df_col.groupby('Sex').agg({
    'Age': ['count', 'mean', 'max'],
    'Amount_spent': ['count', 'min', 'max']
}).sort_values(by= ('Age','count'),ascending= False)
df_grouped.columns.names = ['Variable', 'Mesure']
print(df_grouped)

Variable   Age                  Amount_spent                
Mesure   count       mean   max        count    min      max
Sex                                                         
Woman     1356  46.867257  78.0         1356   2.09  2999.98
Man       1128  46.481383  78.0         1128   2.16  2998.62
Unknow      28  42.357143  72.0           28  52.08  2909.85


**Grouping by multiple columns:** To find total count, maximum and minimum values of Amount_spent by State_names, Gender, and Payment_method then we can pass these columns names under `groupby()` function and add `.agg()` with `count`, `mean`, `max` argument.

In [ ]:
#Group By multiple columns
state_gender_res = df[['State_names','Gender','Payment_method','Amount_spent']].groupby([ 'State_names','Gender', 'Payment_method']).agg(['count', 'min', 'max'])
state_gender_res.head(3)

Amount_spent                 
                                         count     min      max
State_names Gender Payment_method                              
Alabama     Female Card                      6  413.11  2749.37
                   Other                     5  851.25  2789.52
                   PayPal                    5   77.90  2520.85

#### Cross Tabulation (Crosstab)

For creating a simple crosstab between Maritatal_status and Payment_method columns we just use `crosstab()` with both column names.

In [ ]:
pd.crosstab(df.Marital_status, df.Payment_method)

Payment_method  Card  Other  PayPal
Marital_status                     
Married          440    357     667
Single           307    232     497

We can include subtotals by `margins` parameter:

In [ ]:
pd.crosstab(df.Marital_status, df.Payment_method, margins=True, margins_name="Total")

Payment_method  Card  Other  PayPal  Total
Marital_status                            
Married          440    357     667   1464
Single           307    232     497   1036
Total            747    589    1164   2500

If We want a display with percentage than `normalize=True` parameter help

In [ ]:
pd.crosstab(df.Marital_status, df.Payment_method, normalize=True, margins=True, margins_name="Total")

Payment_method    Card   Other  PayPal   Total
Marital_status                                
Married         0.1760  0.1428  0.2668  0.5856
Single          0.1228  0.0928  0.1988  0.4144
Total           0.2988  0.2356  0.4656  1.0000

In this crosstab features, we can pass multiple columns names for grouping and analyzing data. For instance, If we want to see how the `Payment_method` and `Employees_status` are distributed by `Marital_status` then we will pass these columns' names in `crosstab()` function and it will show below.

In [ ]:
pd.crosstab(df.Marital_status, [df.Payment_method, df.Employees_status])

Payment_method        Card                                        Other  \
Employees_status Employees Unemployment self-employed workers Employees   
Marital_status                                                            
Married                185           38            81     136       127   
Single                 117           37            57      96        96   

Payment_method                                         PayPal               \
Employees_status Unemployment self-employed workers Employees Unemployment   
Marital_status                                                               
Married                    39            78     113       265           66   
Single                     17            43      76       178           55   

Payment_method                          
Employees_status self-employed workers  
Marital_status                          
Married                    120     216  
Single                     106     158